### PR TITLE
Rebuild for CUDA 11.8 and CUDA 12

### PR DESCRIPTION
**Pull request by RaulPPelaez**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by RaulPPelaez** [comment]
@conda-forge-admin, please rerender

File: recipe/meta.yaml [hunk 1]
```diff
@@ -40,12 +46,37 @@ requirements:
...
+    {% if cuda_major >= 12 %}
+    - cuda-driver-dev
+    - cuda-cudart-dev
+    - cuda-nvrtc-dev
+    - cuda-nvtx-dev
+    - cuda-nvml-dev
+    - libcublas-dev
+    - libcusparse-dev 
+    - libcufft-dev 
+    - libcurand-dev
```
Comment (jakirkham) [inline recipe/meta.yaml:59]: Think we only want to add these to `requirements/build` when cross-compiling. Have made some suggestions below along these lines

```suggestion
    - cuda-driver-dev                        # [build_platform != target_platform]
    - cuda-cudart-dev                        # [build_platform != target_platform]
    - cuda-nvrtc-dev                         # [build_platform != target_platform]
    - cuda-nvtx-dev                          # [build_platform != target_platform]
    - cuda-nvml-dev                          # [build_platform != target_platform]
    - libcublas-dev                          # [build_platform != target_platform]
    - libcufft-dev                           # [build_platform != target_platform]
    - libcurand-dev                          # [build_platform != target_platform]
    - libcusparse-dev                        # [build_platform != target_platform]
```

File: recipe/meta.yaml [hunk 2]
```diff
@@ -40,12 +46,37 @@ requirements:
...
+    {% if cuda_major >= 12 %}
+    - cuda-driver-dev
+    - cuda-cudart-dev
+    - cuda-nvrtc-dev
+    - cuda-nvtx-dev
+    - cuda-nvml-dev
+    - libcublas-dev
+    - libcusparse-dev 
+    - libcufft-dev
+    - libcurand-dev
```
Comment (jakirkham) [inline recipe/meta.yaml:78]: Went ahead and sorted these. Also removed some extra whitespace that snuck in

```suggestion
    - cuda-driver-dev
    - cuda-cudart-dev
    - cuda-nvrtc-dev
    - cuda-nvtx-dev
    - cuda-nvml-dev
    - libcublas-dev
    - libcufft-dev
    - libcurand-dev
    - libcusparse-dev
```

**Review by jakirkham — COMMENTED** [review]
Thanks Raul! 🙏

Generally this looks good

Added a couple suggestions below

**Comment by RaulPPelaez** [comment]
@conda-forge-admin, please rerender

**Comment by jakirkham** [comment]
@conda-forge-admin , please re-render

File: recipe/meta.yaml [hunk 3]
```diff
@@ -40,12 +46,37 @@ requirements:
...
+    {% if cuda_major >= 12 %}
+    - cuda-driver-dev
+    - cuda-cudart-dev
+    - cuda-nvrtc-dev
+    - cuda-nvtx-dev
+    - cuda-nvml-dev
+    - libcublas-dev
+    - libcufft-dev
+    - libcurand-dev
+    - libcusparse-dev
```
Comment (jakirkham) [inline recipe/meta.yaml:78]:
```suggestion
    - libcusolver-dev
    - libcusparse-dev
```

File: recipe/meta.yaml [hunk 4]
```diff
@@ -40,12 +46,37 @@ requirements:
...
+    {% if cuda_major >= 12 %}
+    - cuda-driver-dev                        # [build_platform != target_platform]
+    - cuda-cudart-dev                        # [build_platform != target_platform]
+    - cuda-nvrtc-dev                         # [build_platform != target_platform]
+    - cuda-nvtx-dev                          # [build_platform != target_platform]
+    - cuda-nvml-dev                          # [build_platform != target_platform]
+    - libcublas-dev                          # [build_platform != target_platform]
+    - libcufft-dev                           # [build_platform != target_platform]
+    - libcurand-dev                          # [build_platform != target_platform]
+    - libcusparse-dev                        # [build_platform != target_platform]
```
Comment (jakirkham) [inline recipe/meta.yaml:59]: Think we are missing cuSOLVER based on this error [from CI]( https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=845152&view=logs&jobId=e9afaa34-1a0f-534e-78fc-fae528ccd915&j=e9afaa34-1a0f-534e-78fc-fae528ccd915&t=48395e60-bc5f-5e45-7455-d404f0c95f3a ):

```C

  /home/conda/feedstock_root/build_artifacts/xformers_1702696922192/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/lib/python3.10/site-packages/torch/include/ATen/cuda/CUDAContext.h:10:10: fatal error: cusolverDn.h: No such file or directory
     10 | #include <cusolverDn.h>
        |          ^~~~~~~~~~~~~~
  compilation terminated.
```

So think we need this change

```suggestion
    - libcusolver-dev                        # [build_platform != target_platform]
    - libcusparse-dev                        # [build_platform != target_platform]
```

**Review by jakirkham — COMMENTED** [review]
Thanks for all your hard work here Raul! 🙏

Pushed a small chance to fix conflicts

Based on the CI results, think we need a small change below

**Comment by jakirkham** [comment]
@conda-forge-admin , please re-render

**Comment by jakirkham** [comment]
@conda-forge-admin , please re-render

**Comment by jakirkham** [comment]
@conda-forge-admin , please restart CI

**Comment by jakirkham** [comment]
@conda-forge-admin , please re-render

**Comment by jakirkham** [comment]
Enabled a swapfile as the builds seem to be resource constrained

**Comment by jakirkham** [comment]
@conda-forge-admin , please re-render

**Comment by jakirkham** [comment]
JFYI we are planning to drop CUDA 11.2 support in conda-forge soon

Please see [this announcement]( https://conda-forge.org/news/2024/03/06/dropping-cuda-112/ ) and this issue ( https://github.com/conda-forge/conda-forge-pinning-feedstock/issues/5339 ) for more details

The smoothest upgrade path would be to add CUDA 11.8. IOW completing this migration PR

Please let us know if you have any questions on next steps

**Comment by jakirkham** [comment]
JFYI CUDA 11.2 is now officially dropped

**Comment by jakirkham** [comment]
@conda-forge-admin , please re-render

**Comment by jakirkham** [comment]
@conda-forge-admin , please re-render

**Comment by jakirkham** [comment]
Looks like there is a checksum mismatch. Snippet below taken from [this CI run]( https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=925799&view=logs&jobId=d0d954b5-f111-5dc4-4d76-03b6c9d0cf7e&j=d0d954b5-f111-5dc4-4d76-03b6c9d0cf7e&t=779a0c34-7b19-5ccb-5da7-fe4579c91392 )

```python
Downloading https://raw.githubusercontent.com/NVIDIA/cutlass/main/LICENSE.txt
Traceback (most recent call last):
  File "/opt/conda/bin/conda-build", line 11, in <module>
    sys.exit(execute())
  File "/opt/conda/lib/python3.10/site-packages/conda_build/cli/main_build.py", line 581, in execute
    api.build(
  File "/opt/conda/lib/python3.10/site-packages/conda_build/api.py", line 250, in build
    return build_tree(
  File "/opt/conda/lib/python3.10/site-packages/conda_build/build.py", line 3762, in build_tree
    packages_from_this = build(
  File "/opt/conda/lib/python3.10/site-packages/conda_build/build.py", line 2547, in build
    try_download(m, no_download_source=False, raise_error=True)
  File "/opt/conda/lib/python3.10/site-packages/conda_build/render.py", line 757, in try_download
    source.provide(metadata)
  File "/opt/conda/lib/python3.10/site-packages/conda_build/source.py", line 1038, in provide
    unpack(
  File "/opt/conda/lib/python3.10/site-packages/conda_build/source.py", line 171, in unpack
    src_path, unhashed_fn = download_to_cache(
  File "/opt/conda/lib/python3.10/site-packages/conda_build/source.py", line 127, in download_to_cache
    raise RuntimeError(
RuntimeError: SHA256 mismatch: '6fc6ad419e4d26571a14c7c4d26b65345865c20dd8686ddcf5e8673f54d3d233' != '5f0ea053230eb92267e9e80ed67ec08899ebc55d093eff4a2273a7b340a48cca'
```

**Comment by h-vetinari** [comment]
I think I fixed the CUDA 11.8 builds in #24 (CUDA 12.0 can come later) - @RaulPPelaez, if you're willing to join as a maintainer here, that would be great. 🙃

**Comment by h-vetinari** [comment]
Picked some fixes from this PR into #24, so this should not be needed anymore. Would still be amazing to have you join as a maintainer here @RaulPPelaez! :)

**Comment by h-vetinari** [comment]
Closing this, but if you're interested in joining the maintenance effort here @RaulPPelaez, please pipe up!

**Pull request closed without merging**